File: src/qrtransfer/protocol.py
"""転送プロトコル v1: フレームの生成と解析、範囲文字列のパースと生成。

フレーム形式（ビッグエンディアン）:
    magic(2) "QZ" | version(1) | type(1) | session_id(4) | seq(4) | total(4) | payload(N) | crc32(4)
CRC32 はオフセット 0 から payload 末尾までを対象とする。

type: 0=META、1=DATA（seq 番目のチャンク）、2=REPAIR（修復用。seq は修復用フレームの番号で、
payload は複数のチャンクの XOR。どのチャンクを重ねたかは repair.py の決まりで session_id と seq から求まる）。
REPAIR を知らない古い受信側は、種類が不明なフレームとして読み飛ばす（DATA だけで従来どおり受信できる）。
"""

from __future__ import annotations

import re
import struct
import zlib
from dataclasses import dataclass
from typing import Iterable

MAGIC = b"QZ"
VERSION = 1
SUPPORTED_VERSIONS = frozenset({1})

TYPE_META = 0
TYPE_DATA = 1
TYPE_REPAIR = 2
FRAME_TYPES = (TYPE_META, TYPE_DATA, TYPE_REPAIR)

HEADER = struct.Struct(">2sBBIII")
HEADER_SIZE = HEADER.size  # 16
CRC = struct.Struct(">I")
CRC_SIZE = CRC.size  # 4
OVERHEAD = HEADER_SIZE + CRC_SIZE  # 20

META_INTERVAL = 20  # DATA 20 枚ごとに META を 1 枚挟む

_U32_MAX = 0xFFFFFFFF


@dataclass(frozen=True)
class Frame:
    type: int
    session_id: int
    seq: int
    total: int
    payload: bytes

    @property
    def is_meta(self) -> bool:
        return self.type == TYPE_META

    @property
    def is_data(self) -> bool:
        return self.type == TYPE_DATA

    @property
    def is_repair(self) -> bool:
        return self.type == TYPE_REPAIR


def build_frame(ftype: int, session_id: int, seq: int, total: int, payload: bytes) -> bytes:
    if ftype not in FRAME_TYPES:
        raise ValueError(f"unknown frame type: {ftype}")
    for name, value in (("session_id", session_id), ("seq", seq), ("total", total)):
        if not 0 <= value <= _U32_MAX:
            raise ValueError(f"{name} out of range: {value}")
    body = HEADER.pack(MAGIC, VERSION, ftype, session_id, seq, total) + bytes(payload)
    return body + CRC.pack(zlib.crc32(body) & _U32_MAX)


def build_meta_frame(session_id: int, total: int, payload: bytes) -> bytes:
    return build_frame(TYPE_META, session_id, 0, total, payload)


def build_data_frame(session_id: int, seq: int, total: int, payload: bytes) -> bytes:
    return build_frame(TYPE_DATA, session_id, seq, total, payload)


def build_repair_frame(session_id: int, index: int, total: int, payload: bytes) -> bytes:
    return build_frame(TYPE_REPAIR, session_id, index, total, payload)


def parse_frame(data: bytes | bytearray | memoryview | None) -> Frame | None:
    """フレームを解析する。不正なフレームは例外を出さずに None を返す（黙って破棄）。"""
    if data is None:
        return None
    data = bytes(data)
    if len(data) < OVERHEAD:
        return None
    magic, version, ftype, session_id, seq, total = HEADER.unpack_from(data, 0)
    if magic != MAGIC or version not in SUPPORTED_VERSIONS:
        return None
    (crc,) = CRC.unpack_from(data, len(data) - CRC_SIZE)
    if zlib.crc32(data[:-CRC_SIZE]) & _U32_MAX != crc:
        return None
    if ftype not in FRAME_TYPES:
        return None
    if ftype == TYPE_DATA and seq >= total:
        return None
    return Frame(ftype, session_id, seq, total, data[HEADER_SIZE:-CRC_SIZE])


# ---------------------------------------------------------------------------
# 範囲文字列 "12,57-60,99"
# ---------------------------------------------------------------------------

_RANGE_ITEM = re.compile(r"^(\d+)(?:\s*-\s*(\d+))?$")
# 全角の数字・記号も受け付ける（手入力のため）
_NORMALIZE = str.maketrans("０１２３４５６７８９，、－ー―‐〜~　", "0123456789,,------ ")


def parse_ranges(text: str, limit: int | None = None) -> set[int]:
    """`"1,3-5,7"` → `{1,3,4,5,7}`。

    limit を指定すると、limit 以上の値を含む入力を拒否する（0 <= n < limit）。
    空文字列（空白のみ）は空集合を返す。不正な入力は ValueError。
    """
    if text is None:
        raise ValueError("input is None")
    text = text.translate(_NORMALIZE).replace("\n", ",").replace("\r", ",")
    if not text.strip():
        return set()
    result: set[int] = set()
    for raw in text.split(","):
        item = raw.strip(" \t")
        if not item:
            raise ValueError(f"empty item in {text!r}")
        m = _RANGE_ITEM.match(item)
        if not m:
            raise ValueError(f"invalid range item: {raw!r}")
        start = int(m.group(1))
        end = int(m.group(2)) if m.group(2) is not None else start
        if end < start:
            raise ValueError(f"reversed range: {raw!r}")
        if limit is not None and end >= limit:
            raise ValueError(f"value out of range (must be < {limit}): {raw!r}")
        if end > _U32_MAX:
            raise ValueError(f"value too large: {raw!r}")
        if end - start > 10_000_000:
            raise ValueError(f"range too large: {raw!r}")
        result.update(range(start, end + 1))
    return result


def iter_ranges(values: Iterable[int]) -> Iterable[tuple[int, int]]:
    """昇順の (start, end) 区間列を返す（end を含む）。"""
    start = prev = None
    for v in sorted(set(values)):
        if start is None:
            start = prev = v
        elif v == prev + 1:
            prev = v
        else:
            yield start, prev
            start = prev = v
    if start is not None:
        yield start, prev


def format_ranges(values: Iterable[int], max_items: int | None = None) -> str:
    """`{1,3,4,5,7}` → `"1,3-5,7"`。

    max_items を指定すると、先頭 max_items 区間だけを出力し、残りがあれば末尾に ",…" を付ける
    （表示用。",…" 付きの文字列は parse_ranges では解析できない）。
    """
    parts: list[str] = []
    for i, (s, e) in enumerate(iter_ranges(values)):
        if max_items is not None and i >= max_items:
            parts.append("…")
            break
        parts.append(str(s) if s == e else f"{s}-{e}")
    return ",".join(parts)

File: web/src/lib/protocol.ts
/**
 * 転送プロトコル v1（Python 版 src/qrtransfer/protocol.py と同じ形式）。
 *
 * フレーム（ビッグエンディアン）:
 *   magic(2) "QZ" | version(1) | type(1) | session_id(4) | seq(4) | total(4) | payload(N) | crc32(4)
 * CRC32 はオフセット 0 から payload 末尾までを対象とする。
 * type: 0=META、1=DATA、2=REPAIR（修復用。seq は修復用フレームの番号。repair.ts を参照）
 */
import { crc32 } from "./crc32";

export const TYPE_META = 0;
export const TYPE_DATA = 1;
export const TYPE_REPAIR = 2;
export const HEADER_SIZE = 16;
export const CRC_SIZE = 4;
export const OVERHEAD = HEADER_SIZE + CRC_SIZE;
const VERSION = 1;
const SUPPORTED_VERSIONS = new Set([VERSION]);
const U32_MAX = 0xffffffff;

/** DATA 20 枚ごとに META を 1 枚挟む（Python 版 protocol.META_INTERVAL） */
export const META_INTERVAL = 20;

export interface Frame {
  type: number;
  sessionId: number;
  seq: number;
  total: number;
  payload: Uint8Array;
}

/** フレームを解析する。不正なフレームは null（黙って破棄）。 */
export function parseFrame(data: Uint8Array | null | undefined): Frame | null {
  if (!data || data.length < OVERHEAD) return null;
  if (data[0] !== 0x51 || data[1] !== 0x5a) return null; // "QZ"
  if (!SUPPORTED_VERSIONS.has(data[2])) return null;
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const crc = view.getUint32(data.length - CRC_SIZE);
  if (crc32(data, 0, data.length - CRC_SIZE) !== crc) return null;
  const type = data[3];
  if (type !== TYPE_META && type !== TYPE_DATA && type !== TYPE_REPAIR) return null;
  const sessionId = view.getUint32(4);
  const seq = view.getUint32(8);
  const total = view.getUint32(12);
  if (type === TYPE_DATA && seq >= total) return null;
  return { type, sessionId, seq, total, payload: data.slice(HEADER_SIZE, data.length - CRC_SIZE) };
}

/** フレームを組み立てる（Python 版 protocol.build_frame と同じバイト列になる）。 */
export function buildFrame(type: number, sessionId: number, seq: number, total: number, payload: Uint8Array): Uint8Array {
  if (type !== TYPE_META && type !== TYPE_DATA && type !== TYPE_REPAIR) throw new RangeError(`unknown frame type: ${type}`);
  for (const [name, v] of [["sessionId", sessionId], ["seq", seq], ["total", total]] as const) {
    if (!(Number.isInteger(v) && v >= 0 && v <= U32_MAX)) throw new RangeError(`${name} out of range: ${v}`);
  }
  const out = new Uint8Array(OVERHEAD + payload.length);
  const view = new DataView(out.buffer);
  out[0] = 0x51; // "Q"
  out[1] = 0x5a; // "Z"
  out[2] = VERSION;
  out[3] = type;
  view.setUint32(4, sessionId);
  view.setUint32(8, seq);
  view.setUint32(12, total);
  out.set(payload, HEADER_SIZE);
  view.setUint32(out.length - CRC_SIZE, crc32(out, 0, out.length - CRC_SIZE));
  return out;
}

// 全角の数字・記号も受け付ける（手入力のため。Python 版 protocol._NORMALIZE と同じ）
const NORMALIZE: Record<string, string> = Object.fromEntries(
  [..."０１２３４５６７８９，、－ー―‐〜~　"].map((c, i) => [c, "0123456789,,------ "[i]]),
);

/**
 * "1,3-5,7" → {1,3,4,5,7}（Python 版 protocol.parse_ranges と同じ）。
 * limit を指定すると limit 以上の値を拒否する。空文字列は空集合。不正な入力は Error。
 */
export function parseRanges(text: string, limit?: number): Set<number> {
  const s = [...text].map((c) => NORMALIZE[c] ?? c).join("").replace(/[\r\n]/g, ",");
  const result = new Set<number>();
  if (!s.trim()) return result;
  for (const raw of s.split(",")) {
    const item = raw.replace(/^[ \t]+|[ \t]+$/g, "");
    if (!item) throw new Error(`空の項目があります: ${text}`);
    const m = /^(\d+)(?:\s*-\s*(\d+))?$/.exec(item);
    if (!m) throw new Error(`番号として読めません: ${raw}`);
    const start = Number(m[1]);
    const end = m[2] !== undefined ? Number(m[2]) : start;
    if (end < start) throw new Error(`範囲が逆です: ${raw}`);
    if (limit !== undefined && end >= limit) throw new Error(`番号が大きすぎます（${limit - 1} まで）: ${raw}`);
    if (end > U32_MAX) throw new Error(`番号が大きすぎます: ${raw}`);
    if (end - start > 10_000_000) throw new Error(`範囲が広すぎます: ${raw}`);
    for (let v = start; v <= end; v++) result.add(v);
  }
  return result;
}

/** 昇順の [start, end] 区間列（end を含む）。 */
export function* iterRanges(values: Iterable<number>): Generator<[number, number]> {
  const sorted = [...new Set(values)].sort((a, b) => a - b);
  let start: number | null = null;
  let prev = 0;
  for (const v of sorted) {
    if (start === null) {
      start = prev = v;
    } else if (v === prev + 1) {
      prev = v;
    } else {
      yield [start, prev];
      start = prev = v;
    }
  }
  if (start !== null) yield [start, prev];
}

/** [1,3,4,5,7] → "1,3-5,7"。maxItems を超える分は ",…" にまとめる（表示用）。 */
export function formatRanges(values: Iterable<number>, maxItems?: number): string {
  const parts: string[] = [];
  let i = 0;
  for (const [s, e] of iterRanges(values)) {
    if (maxItems !== undefined && i >= maxItems) {
      parts.push("…");
      break;
    }
    parts.push(s === e ? String(s) : `${s}-${e}`);
    i++;
  }
  return parts.join(",");
}

export function hex32(n: number): string {
  return (n >>> 0).toString(16).padStart(8, "0");
}
